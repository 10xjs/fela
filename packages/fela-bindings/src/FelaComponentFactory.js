/* @flow */
import { combineRules } from 'fela'

function resolveRule(style, extend) {
  if (extend) {
    return combineRules(style, extend)
  }

  if (typeof style === 'object') {
    return () => style
  }

  return style
}

export default function FelaComponentFactory(
  createElement: Function,
  FelaTheme: Function,
  contextTypes?: Object
): Function {
  function FelaComponent(
    { children, as = 'div', style, extend, ...otherProps },
    { renderer }
  ) {
    // TODO: add warning when no style is not provided
    // TODO: add tests for multiple nested components
    return createElement(FelaTheme, undefined, theme => {
      const rule = resolveRule(style, extend)

      const className = renderer.renderRule(rule, { theme, ...otherProps })

      if (children instanceof Function) {
        return children({
          className,
          theme,
        })
      }

      return createElement(as, { className }, children)
    })
  }

  if (contextTypes) {
    FelaComponent.contextTypes = contextTypes
  }

  return FelaComponent
}
