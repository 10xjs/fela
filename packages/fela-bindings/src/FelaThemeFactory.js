/* @flow */
import deprecate from './_deprecate'

export default function FelaThemeFactory(
  createElement: Function,
  ThemeContext: any
): any {
  function FelaTheme({ children, render }) {
    // TODO: remove with 11.0.0
    deprecate(
      render !== undefined,
      'The `render` prop in FelaTheme is deprecated. It will be removed in react-fela@11.0.0.\nPlease always use `children` instead. See http://fela.js.org/docs/api/bindings/fela-theme'
    )

    const renderFn = children || render

    return createElement(ThemeContext.Consumer, undefined, renderFn)
  }

  return FelaTheme
}
