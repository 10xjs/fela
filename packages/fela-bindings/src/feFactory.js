/* @flow */
// Fe was heavily inspired by glam, both the code as well as the name
// https://github.com/threepointone/glam/blob/master/packages/glam/src/index.js#L83
export default function feFactory(createElement: Function, FelaComponent: any) {
  return function fe(type: any, props: Object = {}, ...children: any) {
    const { css, className, ...otherProps } = props

    if (css) {
      return createElement(FelaComponent, {
        style: css,
        customClass: className,
        render: renderProps =>
          createElement(
            type,
            {
              ...otherProps,
              className: renderProps.className,
            },
            ...children
          ),
      })
    }

    return createElement(type, props, ...children)
  }
}
