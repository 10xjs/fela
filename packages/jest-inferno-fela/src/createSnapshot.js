import { render } from 'inferno'
import { createElement } from 'inferno-create-element'
import { createRenderer } from 'fela'
import { Provider, ThemeProvider } from 'inferno-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  Provider,
  ThemeProvider
)
