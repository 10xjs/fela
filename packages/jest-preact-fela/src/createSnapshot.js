import { createElement, render } from 'preact'
import { createRenderer } from 'fela'
import { RendererProvider, ThemeProvider } from 'preact-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  RendererProvider,
  ThemeProvider
)
