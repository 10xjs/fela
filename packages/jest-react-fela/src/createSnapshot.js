import { createElement } from 'react'
import { render } from 'react-dom'
import { createRenderer } from 'fela'
import { Provider, ThemeProvider } from 'react-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  Provider,
  ThemeProvider
)
