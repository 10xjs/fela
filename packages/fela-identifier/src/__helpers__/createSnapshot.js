import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { render } from 'react-dom'
import { css } from 'js-beautify'

import RendererProviderFactory from '../../../fela-bindings/src/RendererProviderFactory'
import ThemeProviderFactory from '../../../fela-bindings/src/ThemeProviderFactory'
import { THEME_CHANNEL } from '../../../fela-bindings/src/themeChannel'

import renderToString from '../../../fela-tools/src/renderToString'
import createRenderer from '../../../fela/src/createRenderer'

const RendererProvider = RendererProviderFactory(
  Component,
  children => children,
  {
    childContextTypes: { renderer: PropTypes.object },
  }
)

const BaseThemeProvider = ThemeProviderFactory(
  Component,
  children => children,
  {
    propTypes: {
      theme: PropTypes.object.isRequired,
      overwrite: PropTypes.bool,
    },
    childContextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    contextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    defaultProps: {
      overwrite: false,
    },
  }
)

export default function createSnapshot(
  component,
  theme = {},
  renderer = createRenderer(),
  Provider = RendererProvider,
  ThemeProvider = BaseThemeProvider
) {
  const div = document.createElement('div')

  render(
    <Provider renderer={renderer}>
      <ThemeProvider theme={theme}>{component}</ThemeProvider>
    </Provider>,
    div
  )

  return {
    html: div.innerHTML,
    css: css(renderToString(renderer)),
  }
}
