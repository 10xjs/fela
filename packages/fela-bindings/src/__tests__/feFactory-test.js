import 'raf/polyfill'
import React, { createElement, Component } from 'react'
import PropTypes from 'prop-types'

import FelaThemeFactory from '../FelaThemeFactory'
import FelaComponentFactory from '../FelaComponentFactory'
import { THEME_CHANNEL } from '../themeChannel'

import feFactory from '../feFactory'

import createSnapshot from '../__helpers__/createSnapshot'

const FelaTheme = FelaThemeFactory(Component, {
  [THEME_CHANNEL]: PropTypes.object,
})

const FelaComponent = FelaComponentFactory(createElement, FelaTheme, {
  [THEME_CHANNEL]: PropTypes.object,
  renderer: PropTypes.object,
})

const fe = feFactory(createElement, FelaComponent)

describe('Using fe', () => {
  it('should render inline style as CSS', () => {
    const Comp = () =>
      fe(
        'div',
        {
          css: {
            color: 'red',
            ':hover': {
              color: 'blue',
            },
          },
        },
        'Hello'
      )

    expect(createSnapshot(<Comp />)).toMatchSnapshot()
  })

  it('should merge class names', () => {
    const Comp = () =>
      fe(
        'div',
        {
          className: 'Component-button Component',
          css: {
            color: 'red',
            ':hover': {
              color: 'blue',
            },
          },
        },
        'Hello'
      )

    expect(createSnapshot(<Comp />)).toMatchSnapshot()
  })
})
