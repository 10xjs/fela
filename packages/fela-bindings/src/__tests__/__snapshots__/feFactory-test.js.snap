// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using fe should merge class names 1`] = `
Array [
  ".b {
    color: red
}

.a:hover {
    color: blue
}",
  <Provider
    renderer={
        Object {
            "_emitChange": [Function],
            "_renderStyle": [Function],
            "_renderStyleToClassNames": [Function],
            "cache": Object {
              ":hovercolorblue": Object {
                "className": "a",
                "declaration": "color:blue",
                "media": "",
                "pseudo": ":hover",
                "selector": ".a:hover",
                "support": "",
                "type": "RULE",
              },
              "colorred": Object {
                "className": "b",
                "declaration": "color:red",
                "media": "",
                "pseudo": "",
                "selector": ".b",
                "support": "",
                "type": "RULE",
              },
            },
            "clear": [Function],
            "devMode": false,
            "filterClassName": [Function],
            "getNextRuleIdentifier": [Function],
            "keyframePrefixes": Array [
              "-webkit-",
              "-moz-",
              "",
            ],
            "listeners": Array [],
            "mediaQueryOrder": Array [],
            "nodes": Object {},
            "plugins": Array [],
            "renderFont": [Function],
            "renderKeyframe": [Function],
            "renderRule": [Function],
            "renderStatic": [Function],
            "ruleOrder": Array [
              /\\^:link/,
              /\\^:visited/,
              /\\^:hover/,
              /\\^:focus-within/,
              /\\^:focus/,
              /\\^:active/,
            ],
            "selectorPrefix": "",
            "subscribe": [Function],
            "supportQueryOrder": Array [],
            "uniqueKeyframeIdentifier": 0,
            "uniqueRuleIdentifier": 2,
          }
    }
>
    <Comp>
        <FelaComponent
            customClass="foo-bar baz"
            render={[Function]}
            style={
                Object {
                    ":hover": Object {
                      "color": "blue",
                    },
                    "color": "red",
                  }
            }
        >
            <FelaTheme
                render={[Function]}
            >
                <div
                    className="foo-bar baz a b"
                >
                    Hello
                </div>
            </FelaTheme>
        </FelaComponent>
    </Comp>
</Provider>,
]
`;

exports[`Using fe should render inline style as CSS 1`] = `
Array [
  ".b {
    color: red
}

.a:hover {
    color: blue
}",
  <Provider
    renderer={
        Object {
            "_emitChange": [Function],
            "_renderStyle": [Function],
            "_renderStyleToClassNames": [Function],
            "cache": Object {
              ":hovercolorblue": Object {
                "className": "a",
                "declaration": "color:blue",
                "media": "",
                "pseudo": ":hover",
                "selector": ".a:hover",
                "support": "",
                "type": "RULE",
              },
              "colorred": Object {
                "className": "b",
                "declaration": "color:red",
                "media": "",
                "pseudo": "",
                "selector": ".b",
                "support": "",
                "type": "RULE",
              },
            },
            "clear": [Function],
            "devMode": false,
            "filterClassName": [Function],
            "getNextRuleIdentifier": [Function],
            "keyframePrefixes": Array [
              "-webkit-",
              "-moz-",
              "",
            ],
            "listeners": Array [],
            "mediaQueryOrder": Array [],
            "nodes": Object {},
            "plugins": Array [],
            "renderFont": [Function],
            "renderKeyframe": [Function],
            "renderRule": [Function],
            "renderStatic": [Function],
            "ruleOrder": Array [
              /\\^:link/,
              /\\^:visited/,
              /\\^:hover/,
              /\\^:focus-within/,
              /\\^:focus/,
              /\\^:active/,
            ],
            "selectorPrefix": "",
            "subscribe": [Function],
            "supportQueryOrder": Array [],
            "uniqueKeyframeIdentifier": 0,
            "uniqueRuleIdentifier": 2,
          }
    }
>
    <Comp>
        <FelaComponent
            render={[Function]}
            style={
                Object {
                    ":hover": Object {
                      "color": "blue",
                    },
                    "color": "red",
                  }
            }
        >
            <FelaTheme
                render={[Function]}
            >
                <div
                    className="a b"
                >
                    Hello
                </div>
            </FelaTheme>
        </FelaComponent>
    </Comp>
</Provider>,
]
`;
